src/pstore/pstore.c

```diff
@@ -0,0 +1,513 @@
+                                if (pe->contents) {
+                                        dmesg2 = realloc(dmesg, dmesg_size + pe->statbuf.st_size);
+                                        if (dmesg2) {
+                                                dmesg = dmesg2;
+                                                memcpy(&dmesg[dmesg_size], pe->contents, pe->statbuf.st_size);
+                                                dmesg_size += pe->statbuf.st_size;
+                                        }
+                                }
+                        }
+                }
+                if (dmesg) {
+                        (void)write_dmesg(dmesg, dmesg_size, id);
+                        free(dmesg);
+                }
+
+                free_files(&dmesg_files);
+        }
+}
+
+static void move_handler(rule_t *rule, pstore_entry_t *pe) {
+        /* Simply move file out of pstore into archive storage */
+        static pstore_entry_t *move_files = NULL;
+
+        if (pe)
+                LIST_APPEND(dll, move_files, pe);
+        else {
+                /* end of files */
+                LIST_FOREACH(dll, pe, move_files) {
+                        move_file(pe, NULL);
+                }
+                free_files(&move_files);
+        }
+}
+
+static int list_files(pstore_entry_t **list, const char *sourcepath) {
+        DIR *dirp;
+        struct dirent *de;
+        pstore_entry_t *pe;
+        int r = 0;
+
+        errno = 0;
+        dirp = opendir(sourcepath);
+        if (!dirp)
+                return log_error_errno(errno, "Failed to opendir %s: %m", sourcepath);
+
+        do {
+                errno = 0;
+                de = readdir(dirp);
+                if (de && (startswith(de->d_name, ".") == NULL)) {
+                        _cleanup_free_ char *ifd_path;
+                        struct stat statbuf;
+                        int ifd;
+                        ssize_t rr;
+
+                        ifd_path = path_join(sourcepath, de->d_name);
+                        if (!ifd_path)
+                                return log_oom();
+
+                        if (lstat(ifd_path, &statbuf) != 0)
+                                return log_error_errno(errno, "Failed to stat %s: %m", ifd_path);
+
+                        pe = malloc(sizeof(pstore_entry_t));
+                        if (!pe)
+                                return log_oom();
+
+                        pe->direntbuf = *de;
+                        pe->statbuf = statbuf;
+                        pe->is_binary = true;
+                        LIST_INIT(dll, pe);
+                        LIST_APPEND(dll, (*list), pe);
+
+                        /* Now read contents of pstore file */
+                        pe->contents = malloc(pe->statbuf.st_size);
+                        if (!pe->contents)
+                                return log_oom();
+
+                        ifd = open(ifd_path, O_CLOEXEC|O_RDONLY);
```

Comment:
please use openat() for new code, i.e. let's avoid naming races when it's easy

Reply:
in fact, you can just use read_full_file_full() for reading this

Reply:
done